core/src/main/java/org/apache/spark/ExecutorPlugin.java

```diff
@@ -47,6 +48,17 @@
    */
   default void init() {}
 
+  /**
+   * Initialize the executor plugins used to extend the Spark/Dropwizard metrics system.
+   *
+   * <p>Each executor will, during its initialization, invoke this method on each
+   * plugin provided in the spark.executor.metrics.plugins configuration.</p>
+   *
+   * <p>Plugins should register the data sources using the Dropwizard/codahale API</p>
+   *
+   */
+  default void init(MetricRegistry sourceMetricsRegistry) {}
```

Comment:
I prefer `initMetrics`, and explain that it will be called after `init()`.

Reply:
Thanks @vanzin for looking at this. I'll be interested to know about your use case for using this (executor plugins for extending the metrics system) . BTW I take the occasion to add that over the summer we have used this code a few times for workload and performance measurements/tests, and found it quite useful, in particular in the context of measuring I/O access time with some custom plugins we worte ( https://github.com/cerndb/SparkExecutorPlugins ) + custom I/O instrumentation for S3, HDFS. I have been thinking also at adding some additional instrumentation for CPU counters or network metrics, but not yet worked on that.
I agree that using one config for "normal" plugins and for metrics plugins would reduce complexity and in general be preferrable. I'll appreciate a few more deatils on your proposed changes. I guess what could be a very simple way to merge the two plugin types, is just to pass sourceMetricsregistry to all plugins init code. This would be a breaking change from 2.4, but maybe acceptable for Spark 3.0? I guess there are just a few people using executor plugins in their current form? /cc @squito

Reply:
I just want to add a new optional metrics source to Spark, and plugging into this functionality seemed nicer than having ad-hoc code. It would also mean this API would be used also inside of Spark.

>  just to pass sourceMetricsregistry to all plugins init code

Either is fine, it's a `DeveloperApi` so we can make those changes. But, for future proofing, I'd add a level of indirection - add a `ExecutorPluginContext` class that holds references to things provided to the init method (such as the metrics registry). Then adding things to that class does not break the API.

Reply:
sorry I forgot to come back to look at this pr.  I agree with marcelo's comment.

I think its fine to the break the api from 2.4.  If we want, you could even still support old plugins by using reflection.

(or even version the api, but that seems like overkill at this point)